makefile

```diff
@@ -62,6 +62,7 @@ test_all: venv wait_db test lint
 .PHONY: test_all
 
 doc: venv clean_docs
+	python3 docs/source/utils/patching.py
```

Comment:
I think that patching library files like that is a very poor practice (as noted in the documentation for that file).

------

I suggest three solutions:

1. Mention as much as possible that running `make doc` is unsafe.
2. Unpatch files at the end of `doc`. This would require storing backups of patched files.
3. Patch only in runtime. Here is an example of patching `os.listdir` which is used in `sphinx.ext.apidoc.is_packagedir`:

```python
import os

os.listdir = lambda *_, **__: ("__init__.abc", )


import sphinx.ext.apidoc as module


assert module.is_initpy("__init__.py")  # uses `PY_SUFFIXES` to determine if a file is __init__
assert not module.is_packagedir(dirname="/")  # uses `os.listdir` and `is_initpy` to determine if there are __init__ files in a dir

module.PY_SUFFIXES = (".abc", )

assert not module.is_initpy("__init__.py")
assert module.is_packagedir(dirname="/")

```

Using this approach would require creating a `.py` file that does
```commandline
sphinx-apidoc -e -E -f -o docs/source/apiref dff
sphinx-build -M clean docs/source docs/build
source <(cat .env_file | sed 's/=/=/' | sed 's/^/export /') && export DISABLE_INTERACTIVE_MODE=1 && sphinx-build -b html -W --keep-going docs/source docs/build
```

`sphinx-apidoc` calls `sphinx.ext.apidoc.main`; `sphinx-build` calls `sphinx.cmd.build.main`.

------

I think the third solution is the best one as it never edits any library files.
The resulting file would look like this:

`doc.py`
```
import sphinx.ext.autosummary
sphinx.ext.autosummary.extract_summary = ...
import nbsphinx
nbsphinx.depart_gallery_html = ...

import sphinx.ext.apidoc
import sphinx.cmd.build

sphinx.ext.apidoc.main(...)
sphinx.cmd.build.main(...)
```

Note: this may not work if either `sphinx.ext.apidoc` or `sphinx.cmd.build` is imported in `sphinx.ext.autosummary` or `nbsphinx` in which case [`importlib.reload`](https://docs.python.org/3/library/importlib.html#importlib.reload) might help.

Reply:
I prefer to push it to another PR, The idea of @RLKRo is awesome. Let's merge this PR and change the patch method to third one of @RLKRo 's ones

Reply:
@kudep
Do we plan to release a new library version between this PR and the one that is supposed to fix this issue?

If so I think we should implement at least on of the proposed solutions (e.g. mention in `CONTRIBUTING.md` that `make doc` is supposed to be run only inside a virtual environment).

If we do not plan to release a new version of the library, then I propose the following solution:

We consider this PR ready and complete. Create a new branch (and a new PR into dev) from here in which the fix is implemented. Then merge both PRs.